New protocol documentation. (#380)
LiskHQ/lisk-docs: modules/ROOT/pages/2-transactions.adoc

= Transactions

> In Lisk, the state of the blockchain is updated by transactions included in a link:3-blocks.adoc[block]. Every account can issue a transaction, with the corresponding signature and data. There is a set of different transaction types and each of these types defines a specific logic for modifying accounts depending on the transaction data. In the table below, a summary of these transaction types and the purpose each of them serve is shown. 

Author: ```suggestion
In Lisk, the state of accounts is updated by transactions included in a link:3-blocks.adoc[block]. Every account can issue a transaction, with the corresponding signature and data. There is a set of different transaction types and each of these types defines a specific logic for modifying accounts depending on the transaction data. In the table below, a summary of these transaction types and the purpose each of them serve is shown. 
```

Author: This would be more inline with the intro and other parts of the protocol documentation.

Reviewer: done

=== ID

In Lisk, every transaction is associated with a transaction ID, which uniquely identifies a transaction in the blockchain. The transaction ID is the SHA-256 hash of the serialized transaction. 

=== Minimum Balance

A <<Balance Transfer>> transaction must make the balance of the receiving account equal to or greater than 5 × 10^6^ (0.05 LSK in Lisk Mainnet). Subsequently, any transaction sent from an account is only valid if the resulting balance of the sender account is at least 5 × 10^6^. This constraint exists to account for the cost of creating and storing accounts in Lisk.

== Transaction Types

> The 6 transactions types available in the Lisk protocol are listed in the table displayed above. Each one induces a concrete operation on the blockchain state and is defined by a unique schema and transaction logic. The type-specific data is given in the `asset` property previously mentioned. The key points and useful information for each of these transactions are commented in the following subsections.

Author: ```suggestion
The 6 transactions types available in the Lisk protocol are listed in the table displayed above. Each one induces a concrete operation on accounts and is defined by a unique schema and transaction logic. The type-specific data is given in the `asset` property previously mentioned. The key points and useful information for each of these transactions are commented in the following subsections.
```

Author: We talk about operations on the accounts state rather than blockchain state in the other sections.

Reviewer: done

For example, for a <<Balance Transfer>> transaction with a size of 133 bytes, the minimum fee is 10^-5^ LSK/byte × 133 bytes = 0.00133 LSK. In the case of <<Delegate Registration>> transaction with a size of 124 bytes, then the__ __minimum fee of the transaction is 10 LSK + 10^-5^ LSK/byte × 124 bytes = 10.00124 LSK.

This required minimum fee paid by every transaction is burned, i.e. it is not assigned to any account balance. The remaining fee on top of it is assigned to the delegate forging the block in which the transaction is included. This implies that under normal circumstances delegates give priority to transactions with higher remaining fee.

==== `senderPublicKey`

> The public key of the account issuing the transaction. Note that, this public key does not necessarily sign the transaction. For simplicity, the account issuing the transaction is called sender account in the rest of document.

Reviewer: Going for the first option for simplicity and being more generic

==== `asset` 

> This property contains the data logic specific to the type of the transaction. For each `type` value, there exist a unique `asset` property. The next section describes the `asset` property of the different transaction types.

Reviewer: reformulated the 'asset' sentence to make it more correct.

==== `signatures`

> An array with the signatures of the transaction. If the account associated to `senderPublicKey` does not have the link:1-accounts.adoc#keys[`keys` property], the object containing the five properties introduced above has to be signed by the private key corresponding to the `senderPublicKey`. Otherwise, the signing process has to be repeated for each of the private keys associated with the applicable public keys specified in  the `keys` property of the account. In the link:6-appendix.adoc#signature-scheme[Appendix] section a detailed of the signing process is given.

Reviewer: that is why I said "applicable public keys". This is, for the public keys that apply to the specific case. I think that conveys the message better than saying "some public keys".

> An array with the signatures of the transaction. If the account associated to `senderPublicKey` does not have the link:1-accounts.adoc#keys[`keys` property], the object containing the five properties introduced above has to be signed by the private key corresponding to the `senderPublicKey`. Otherwise, the signing process has to be repeated for each of the private keys associated with the applicable public keys specified in  the `keys` property of the account. In the link:6-appendix.adoc#signature-scheme[Appendix] section a detailed of the signing process is given.

Reviewer 2: But I would say that every public key from the keys property is applicable in every case. I'm not sure what you mean with "apply to the specific case".

> An array with the signatures of the transaction. If the account associated to `senderPublicKey` does not have the link:1-accounts.adoc#keys[`keys` property], the object containing the five properties introduced above has to be signed by the private key corresponding to the `senderPublicKey`. Otherwise, the signing process has to be repeated for each of the private keys associated with the applicable public keys specified in  the `keys` property of the account. In the link:6-appendix.adoc#signature-scheme[Appendix] section a detailed of the signing process is given.

Reviewer: changed to undefined set as suggested for the multisig case.

==== ID

> In Lisk, every transaction is associated with a transaction ID, which uniquely identifies a transaction in the blockchain. The transaction ID is the SHA-256 hash of the serialized  transaction.  +

Reviewer: note that ID here is not the specific name of the property/variable/object in the protocol but just the name of the section

> In Lisk, every transaction is associated with a transaction ID, which uniquely identifies a transaction in the blockchain. The transaction ID is the SHA-256 hash of the serialized  transaction.  +

Reviewer 2: I was referring to "transaction ID" in line 53. By reading the all serialization LIPs, I also got the impression that "transation ID" and "block ID" are preferred over "transactionID" and "blockID". Was it decided at some point we use this way?

> In Lisk, every transaction is associated with a transaction ID, which uniquely identifies a transaction in the blockchain. The transaction ID is the SHA-256 hash of the serialized  transaction.  +

Reviewer: I don't recall the exact moment and places, but I remember Jan making the point a couple of times that transaction/block/etc IDs should be written separately because it is a concept and not a property name per se. The same one would say about block header or transaction signature. Probably this was discussed on one of these doc style meetings before Christmas.

> Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:6-network.adoc#transaction-pool[transaction pool] and eventually in a link:3-blocks.adoc[block]. 

Reviewer: the second sentence tries to put in perspective (respect to other parts of the protocol/application) what happens after all these things are calculated. I could remove it if needed, but I think it is not a big deal and clarifies the basics for a novel reader.

> Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:6-network.adoc#transaction-pool[transaction pool] and eventually in a link:3-blocks.adoc[block]. 

Reviewer 2: The ID is not even strictly needed for transmission. Me removed the ID properties from the transaction schema in the "Remove redundant properties" LIP. I'm not even sure how the API works, i.e., if a front end tool is sending a transaction with or without ID to a node. But this is anyway just an implementation detail and work in theory without transactionID.

> Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:6-network.adoc#transaction-pool[transaction pool] and eventually in a link:3-blocks.adoc[block]. 

Reviewer: would you just remove the whole sentence? or do you think it is ok to not be totally precise here to connect this with the transaction pool concept?

> Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:6-network.adoc#transaction-pool[transaction pool] and eventually in a link:3-blocks.adoc[block]. 

Reviewer 2: I would remove it.

> Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:6-network.adoc#transaction-pool[transaction pool] and eventually in a link:3-blocks.adoc[block]. 

Reviewer: removed

> ==== Minimum Balance

Reviewer: note that the name of the section containing this one is "Transaction Properties, Transaction ID and Minimum Balance", so in that regard it should not be a big issue. Given that, I also think that the minimum balance concept does not fit in any section properly as of now.

> ==== Minimum Balance

Reviewer 2: please see my other comment: https://github.com/LiskHQ/lisk-docs/pull/380/files/76ba3fcd3be096b1011366dadf9929203b3fc418..6bf96e0851528c0286929147030649c4ff055327#r369615581

> ==== Minimum Balance

Reviewer: addressed in that comment as you said.

==== Unlock Vote

This transaction [#index-unlocked-2]#*unlocks*# the LSK tokens specified in `amount` that were previously unvoted for the delegate in `delegateAddress` by vote transaction at the height given in the property `unvoteHeight`. This transaction is only valid if it is issued after the unlocking period has been completed since `unvoteHeight`. For a regular vote the unlocking period is 2000 blocks (around 5 hours). For self-votes, i.e. if the `delegateAddress` property of the transaction is equal to the account link:1-accounts.adoc#address[address], this period is 260,000 blocks (around 30 days).

==== Multisignature 

> This transaction registers the sender account as a multisignature account. The set of [#index-mandatory-2]#*mandatory*# keys needs to be specified in `mandatoryKeys` whereas the set of [#index-optional-2]#*optional*# keys have to be specified in `optionalKeys`. The total number of needed keys for every future outgoing transaction from the account is given in `numberOfSignatures`. Once this transaction is included in a block, every transaction from this account has to be signed by the applicable set of signatures.

Reviewer 2: "signed by ... set of signatures" -> "signed by ... set of private keys"?

> This transaction registers the sender account as a multisignature account. The set of [#index-mandatory-2]#*mandatory*# keys needs to be specified in `mandatoryKeys` whereas the set of [#index-optional-2]#*optional*# keys have to be specified in `optionalKeys`. The total number of needed keys for every future outgoing transaction from the account is given in `numberOfSignatures`. Once this transaction is included in a block, every transaction from this account has to be signed by the applicable set of signatures.

Reviewer: good catch, changed

> ==== PoM 

Reviewer 2: @IkerAlus you don't agree? If not, it's ok.

> ==== PoM 

Reviewer: added

|*Vote*|Submit or remove vote(s) for delegates
|*Unlock Vote*|Unlock locked tokens
|*Multisignature*|Register a multisignature in the account
|*PoM*|Submit a Proof-of-Misbehavior of a delegate
|===

== Transaction Properties, Transaction ID and Minimum Balance

image::../assets/images/InfographicsV1/Infographic7.png[TxDiagram]

> Every transaction object in the Lisk protocol, regardless of the type, has a common set of properties. These properties are shown in the subsections below.

Reviewer 2: Despite the graphic, I think that it is not obvious which properties are the common properties, because all following headers, including the ones for ID and Minimum Balance, have level 3.

I think it would be good add an additional header, "Common Properties" or so, of level 3 before this line or before the graphic, and then change header level of these properties to 4.

Reviewer: going point, changed

For example, for a <<Balance Transfer>> transaction with a size of 133 bytes, the minimum fee is 10^-5^ LSK/byte × 133 bytes = 0.00133 LSK. In the case of <<Delegate Registration>> transaction with a size of 124 bytes, the minimum fee of the transaction is 10 LSK + 10^-5^ LSK/byte × 124 bytes = 10.00124 LSK.

This required minimum fee paid by every transaction is burned, in other words, it is not assigned to any account balance. The remaining fee on top of this is assigned to the delegate forging the block in which the transaction is included. This implies that under normal circumstances delegates give priority to transactions with higher remaining fees.

=== `senderPublicKey`

The public key of the account issuing the transaction. Note that this public key does not necessarily sign the transaction. For simplicity, the account issuing the transaction is called the sender account in this document.

=== `asset` 

> This property contains the data specific to the type of the transaction. For each `type` value, there exist a unique `asset` property. The next section describes the different transaction types respect to the `asset` property of each of them.

Reviewer 2: > The next section describes the different transaction types respect to the `asset` property of each of them.

But the next section is `signatures` and not "Transaction Types". Or is a section for you always starting with a level 2 header? If so, I would reformulate it to "The section Transaction Types describes ..." and add a link.

Reviewer: done!

> This property contains the data specific to the type of the transaction. For each `type` value, there exist a unique `asset` property. The next section describes the different transaction types respect to the `asset` property of each of them.

Reviewer 2: ```suggestion
This property contains the data specific to the type of the transaction. For each `type` value, there exist a unique `asset` property. The next section describes the different transaction types with respect to the `asset` property of each of them.
```

Reviewer: done

=== ID

In Lisk, every transaction is associated with a transaction ID, which uniquely identifies a transaction in the blockchain. The transaction ID is the SHA-256 hash of the serialized transaction.  
Once the transaction has all the required signatures and ID, it is ready to be broadcast to the network so that it can be included into the link:5-network.adoc#transaction-pool[transaction pool] and eventually in to a link:3-blocks.adoc[block]. 

=== Minimum Balance

A <<Balance Transfer>> transaction must make the balance of the receiving account equal to or greater than 0.05 LSK tokens. Subsequently, any transaction sent from an account is only valid if the resulting balance of the sender account is at least 0.05 LSK. This constraint exists to account for the cost of creating and storing accounts in Lisk.

== Transaction Types

> The 6 transactions types available in the Lisk protocol are listed in the table displayed above. Each one induces a concrete operation on the blockchain state and is defined by a unique schema and transaction logic. The type-specific data is given in the `asset` property previously mentioned. The key points and useful information for each of these transaction are commented in the following subsections.

Reviewer 2: ```suggestion
The 6 transactions types available in the Lisk protocol are listed in the table displayed above. Each one induces a concrete operation on the blockchain state and is defined by a unique schema and transaction logic. The type-specific data is given in the `asset` property previously mentioned. The key points and useful information for each of these transactions are commented in the following subsections.
```
